core/src/main/java/org/lflang/analyses/statespace/StateSpaceExplorer.java

```diff
@@ -0,0 +1,329 @@
+/** Explores the state space of an LF program. */
+package org.lflang.analyses.statespace;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
+import org.eclipse.xtext.xbase.lib.Exceptions;
+import org.lflang.TimeUnit;
+import org.lflang.TimeValue;
+import org.lflang.generator.ActionInstance;
+import org.lflang.generator.PortInstance;
+import org.lflang.generator.ReactionInstance;
+import org.lflang.generator.ReactorInstance;
+import org.lflang.generator.RuntimeRange;
+import org.lflang.generator.SendRange;
+import org.lflang.generator.TimerInstance;
+import org.lflang.generator.TriggerInstance;
+import org.lflang.lf.Expression;
+import org.lflang.lf.Time;
+import org.lflang.lf.Variable;
+
+public class StateSpaceExplorer {
+
+  // Instantiate an empty state space diagram.
+  public StateSpaceDiagram diagram = new StateSpaceDiagram();
+
+  // Indicate whether a back loop is found in the state space.
+  // A back loop suggests periodic behavior.
+  public boolean loopFound = false;
+
+  /**
+   * Instantiate a global event queue. We will use this event queue to symbolically simulate the
+   * logical timeline. This simulation is also valid for runtime implementations that are federated
+   * or relax global barrier synchronization, since an LF program defines a unique logical timeline
+   * (assuming all reactions behave _consistently_ throughout the execution).
+   */
+  public EventQueue eventQ = new EventQueue();
+
+  /** The main reactor instance based on which the state space is explored. */
+  public ReactorInstance main;
+
+  // Constructor
+  public StateSpaceExplorer(ReactorInstance main) {
+    this.main = main;
+  }
+
+  /** Recursively add the first events to the event queue. */
+  public void addInitialEvents(ReactorInstance reactor) {
+    // Add the startup trigger, if exists.
+    var startup = reactor.getStartupTrigger();
+    if (startup != null) eventQ.add(new Event(startup, new Tag(0, 0, false)));
+
+    // Add the initial timer firings, if exist.
+    for (TimerInstance timer : reactor.timers) {
+      eventQ.add(new Event(timer, new Tag(timer.getOffset().toNanoSeconds(), 0, false)));
+    }
+
+    // Recursion
+    for (var child : reactor.children) {
+      addInitialEvents(child);
+    }
+  }
+
+  /**
+   * Explore the state space and populate the state space diagram until the specified horizon (i.e.
+   * the end tag) is reached OR until the event queue is empty.
+   *
+   * <p>As an optimization, if findLoop is true, the algorithm tries to find a loop in the state
+   * space during exploration. If a loop is found (i.e. a previously encountered state is reached
+   * again) during exploration, the function returns early.
+   *
+   * <p>TODOs: 1. Handle action with 0 min delay. 2. Check if zero-delay connection works.
+   */
+  public void explore(Tag horizon, boolean findLoop) {
```

Comment:
The cyclomatic complexity of this method is probably around 24 based on the number of conditional branches and loops that it has, and 10 is considered a reasonable upper limit. Consider breaking this into smaller methods that each have documentation. Extremely long methods with high cyclomatic complexity and inline comments instead of method-level comments are an antipattern in our code base.

Reply:
Yes, I see your concern @petervdonovan, and I agree this is not the best practice currently. Breaking this method apart could be non-trivial, however. If it is okay with you, can I address this issue in a future PR, maybe in #1830? PR #1830 reuses this infrastructure to do scheduling, and there will be a rather big refactoring to the `statespace` directory based on @oowekyala's comments. Perhaps it's better to do everything in one go, so there won't be too many conflicts.

Reply:
This part of the PR is most concerning, as it also has a bunch of FIXMEs and debugging code in it. If you plan to address this in #1830 soon, I'm OK with merging this as-is, but it need to be a priority in further development. Perhaps leave a comment in the class description stating that this code is somewhat experimental.